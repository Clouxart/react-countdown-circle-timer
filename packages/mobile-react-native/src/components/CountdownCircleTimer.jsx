import React from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, View } from 'react-native'
import Svg, { Path, Defs, LinearGradient } from 'react-native-svg'
import {
  linearEase,
  getWrapperStyle,
  getTimeStyle,
  getStroke,
  colorsValidator,
  useMemoizedProps,
} from '@countdown-circle-timer/shared'

const CountdownCircleTimer = (props) => {
  const {
    size,
    strokeWidth,
    trailColor,
    duration,
    isPlaying,
    colors,
    strokeLinecap,
    children,
    isLinearGradient,
    gradientUniqueKey,
    onComplete,
    ariaLabel,
    renderAriaTime,
    initialRemainingTime,
  } = props

  const {
    path,
    pathLength,
    durationMilliseconds,
    startAt,
    normalizedColors,
    gradientId,
  } = useMemoizedProps({
    size,
    strokeWidth,
    duration,
    initialRemainingTime,
    colors,
    isLinearGradient,
    gradientUniqueKey,
  })

  const elapsedTime = 0
  const stroke = getStroke(normalizedColors, elapsedTime)
  const strokeDashoffset = linearEase(
    elapsedTime,
    0,
    pathLength,
    durationMilliseconds
  ).toFixed(3)
  const timeProps = {
    remainingTime: Math.ceil((durationMilliseconds - elapsedTime) / 1000),
    elapsedTime,
  }

  const styles = StyleSheet.create({
    wrapper: getWrapperStyle(size),
    time: getTimeStyle(stroke, size),
  })

  return (
    <View
      style={styles.wrapper}
      accessible={true}
      accessibilityLabel={ariaLabel}
    >
      <Svg width={size} height={size}>
        {isLinearGradient && (
          <Defs>
            <LinearGradient id={gradientId} x1="100%" y1="0%" x2="0%" y2="0%">
              {normalizedColors.map((color) => (
                <stop {...color.gradient} />
              ))}
            </LinearGradient>
          </Defs>
        )}
        <Path
          fill="none"
          strokeWidth={strokeWidth}
          stroke={trailColor}
          d={path}
        />
        <Path
          fill="none"
          stroke={isLinearGradient ? `url(#${gradientId})` : stroke}
          d={path}
          strokeLinecap={strokeLinecap}
          strokeWidth={strokeWidth}
          strokeDasharray={pathLength}
          strokeDashoffset={strokeDashoffset}
        />
      </Svg>
      {children !== null && (
        <View accessibilityElementsHidden={true} style={styles.time}>
          {React.isValidElement(children)
            ? React.cloneElement(React.Children.only(children), timeProps)
            : children(timeProps)}
        </View>
      )}
      {typeof renderAriaTime === 'function' && (
        <View accessibilityRole="timer" accessibilityLiveRegion="assertive">
          {renderAriaTime(timeProps)}
        </View>
      )}
    </View>
  )
}

CountdownCircleTimer.propTypes = {
  duration: PropTypes.number.isRequired,
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  colors: PropTypes.arrayOf(PropTypes.arrayOf(colorsValidator).isRequired)
    .isRequired,
  size: PropTypes.number,
  strokeWidth: PropTypes.number,
  trailColor: PropTypes.string,
  isPlaying: PropTypes.bool,
  strokeLinecap: PropTypes.oneOf(['round', 'square']),
  isLinearGradient: PropTypes.bool,
  gradientUniqueKey: PropTypes.string,
  onComplete: PropTypes.func,
  ariaLabel: PropTypes.string,
  renderAriaTime: PropTypes.func,
  initialRemainingTime: PropTypes.number,
}

CountdownCircleTimer.defaultProps = {
  size: 180,
  strokeWidth: 12,
  trailColor: '#d9d9d9',
  isPlaying: false,
  strokeLinecap: 'round',
  isLinearGradient: false,
  ariaLabel: 'Countdown timer',
  children: null,
}

CountdownCircleTimer.displayName = 'CountdownCircleTimer'

export { CountdownCircleTimer }
