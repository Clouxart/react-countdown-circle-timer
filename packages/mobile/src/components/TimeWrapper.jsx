import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { Text, Animated, StyleSheet } from 'react-native'
import {
  countdownCircleTimerProps,
  timeStyle,
} from '@countdown-circle-timer/shared'

const styles = StyleSheet.create({
  time: timeStyle,
  ariaTime: { height: 0, opacity: 0 },
})

const TimeWrapper = (props) => {
  const {
    children,
    animatedElapsedTime,
    durationMilliseconds,
    renderAriaTime,
    animatedColor,
  } = props

  const [timeProps, setTimeProps] = useState({
    animatedElapsedTime: 0,
    remainingTime: durationMilliseconds / 1000,
    animatedColor,
  })

  useEffect(() => {
    const animatedListenerId = animatedElapsedTime.addListener(({ value }) => {
      setTimeProps({
        remainingTime: Math.ceil((durationMilliseconds - value) / 1000),
        elapsedTime: value,
        animatedColor,
      })
    })

    return () => {
      animatedElapsedTime.removeListener(animatedListenerId)
    }
  }, [animatedElapsedTime, animatedColor, durationMilliseconds])

  return (
    <>
      {children !== null && (
        <Animated.View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          style={styles.time}
        >
          {React.isValidElement(children)
            ? React.cloneElement(React.Children.only(children), timeProps)
            : children(timeProps)}
        </Animated.View>
      )}
      {typeof renderAriaTime === 'function' && (
        <Text
          accessibilityRole="timer"
          accessibilityLiveRegion="assertive"
          importantForAccessibility="yes"
          style={styles.ariaTime}
        >
          {renderAriaTime(timeProps)}
        </Text>
      )}
    </>
  )
}

TimeWrapper.propTypes = {
  durationMilliseconds: PropTypes.number.isRequired,
  animatedElapsedTime: PropTypes.object.isRequired,
  // when there is a single color we just past the color string
  // when there are more colors it is an interpolate object
  animatedColor: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  children: countdownCircleTimerProps.children,
  renderAriaTime: countdownCircleTimerProps.renderAriaTime,
}

export { TimeWrapper }
