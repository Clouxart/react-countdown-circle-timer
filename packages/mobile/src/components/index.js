export { CircleTimer } from './CircleTimer'
export { TimeWrapper } from './TimeWrapper'
