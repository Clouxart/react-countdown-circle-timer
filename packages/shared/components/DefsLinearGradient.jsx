import React from 'react'
import PropTypes from 'prop-types'
import { countdownCircleTimerProps } from '..'

const getStopProps = (colors) => {
  if (colors.length === 1) {
    return [{ offset: 1, stopColor: colors[0][0], key: 0 }]
  }

  const colorsLength = colors.length
  let currentDuration = 0
  return [
    { offset: 0, stopColor: colors[0][0], key: 0 },
    ...colors.map(([stopColor, duration], index) => {
      const isLastColor = colorsLength === index + 1
      currentDuration += duration

      return {
        offset: isLastColor ? 1 : currentDuration,
        stopColor,
        key: index + 1,
      }
    }),
  ]
}

const DefsLinearGradient = ({
  gradientId,
  colors,
  defs: Defs = 'defs',
  linearGradient: LinearGradient = 'linearGradient',
  stop: Stop = 'stop',
}) => (
  <Defs>
    <LinearGradient id={gradientId} x1="1" y1="0" x2="0" y2="0">
      {getStopProps(colors).map((gradient) => (
        <Stop {...gradient} />
      ))}
    </LinearGradient>
  </Defs>
)

DefsLinearGradient.propTypes = {
  gradientId: PropTypes.string.isRequired,
  colors: countdownCircleTimerProps.colors,
  defs: PropTypes.node,
  linearGradient: PropTypes.node,
  stop: PropTypes.node,
}

export { DefsLinearGradient }
