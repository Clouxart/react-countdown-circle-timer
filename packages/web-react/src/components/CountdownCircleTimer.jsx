import React, { useMemo } from 'react'
import PropTypes from 'prop-types'
import { useElapsedTime } from 'use-elapsed-time'
import {
  linearEase,
  getWrapperStyle,
  getTimeStyle,
  svgStyle,
  getStroke,
  colorsValidator,
  visuallyHidden,
  useMemoizedProps,
} from '@countdown-circle-timer/shared'

const CountdownCircleTimer = (props) => {
  const {
    size,
    strokeWidth,
    trailColor,
    durationSeconds,
    isPlaying,
    colors,
    strokeLinecap,
    renderTime,
    isLinearGradient,
    gradientUniqueKey,
    onComplete,
    ariaLabel,
    renderAriaTime,
    initialRemainingTime,
    startAt: startAtSeconds,
  } = props

  const {
    path,
    pathLength,
    durationMilliseconds,
    startAt,
    normalizedColors,
    gradientId,
  } = useMemoizedProps({
    size,
    strokeWidth,
    durationSeconds,
    initialRemainingTime,
    startAtSeconds,
    colors,
    isLinearGradient,
    gradientUniqueKey,
  })

  const elapsedTime = useElapsedTime(isPlaying, {
    durationMilliseconds,
    onComplete,
    startAt,
  })
  const strokeDashoffset = linearEase(
    elapsedTime,
    0,
    pathLength,
    durationMilliseconds
  ).toFixed(3)
  const stroke = getStroke(normalizedColors, elapsedTime)
  const remainingTime = Math.ceil((durationMilliseconds - elapsedTime) / 1000)

  return (
    <div style={getWrapperStyle(size)} aria-label={ariaLabel}>
      <svg
        width={size}
        height={size}
        style={svgStyle}
        xmlns="http://www.w3.org/2000/svg"
      >
        {isLinearGradient && (
          <defs>
            <linearGradient id={gradientId} x1="100%" y1="0%" x2="0%" y2="0%">
              {normalizedColors.map((color) => (
                <stop {...color.gradient} />
              ))}
            </linearGradient>
          </defs>
        )}
        <path
          fill="none"
          strokeWidth={strokeWidth}
          stroke={trailColor}
          d={path}
        />
        <path
          fill="none"
          stroke={isLinearGradient ? `url(#${gradientId})` : stroke}
          d={path}
          strokeLinecap={strokeLinecap}
          strokeWidth={strokeWidth}
          strokeDasharray={pathLength}
          strokeDashoffset={strokeDashoffset}
        />
      </svg>
      {typeof renderTime === 'function' && (
        <div aria-hidden="true" style={getTimeStyle(stroke, size)}>
          {renderTime(remainingTime, elapsedTime, isPlaying)}
        </div>
      )}
      {typeof renderAriaTime === 'function' && (
        <div role="timer" aria-live="assertive" style={visuallyHidden}>
          {renderAriaTime(remainingTime, elapsedTime, isPlaying)}
        </div>
      )}
    </div>
  )
}

CountdownCircleTimer.propTypes = {
  durationSeconds: PropTypes.number.isRequired,
  colors: PropTypes.arrayOf(PropTypes.arrayOf(colorsValidator).isRequired)
    .isRequired,
  size: PropTypes.number,
  strokeWidth: PropTypes.number,
  trailColor: PropTypes.string,
  isPlaying: PropTypes.bool,
  strokeLinecap: PropTypes.oneOf(['round', 'square']),
  renderTime: PropTypes.func,
  isLinearGradient: PropTypes.bool,
  gradientUniqueKey: PropTypes.string,
  onComplete: PropTypes.func,
  ariaLabel: PropTypes.string,
  renderAriaTime: PropTypes.func,
  initialRemainingTime: PropTypes.number,
  startAt: PropTypes.number, // To be removed in next major release v.2 use initialRemainingTime instead
}

CountdownCircleTimer.defaultProps = {
  size: 180,
  strokeWidth: 12,
  trailColor: '#d9d9d9',
  isPlaying: false,
  strokeLinecap: 'round',
  isLinearGradient: false,
  ariaLabel: 'Countdown timer',
  startAt: 0,
}

CountdownCircleTimer.displayName = 'CountdownCircleTimer'

export { CountdownCircleTimer }
