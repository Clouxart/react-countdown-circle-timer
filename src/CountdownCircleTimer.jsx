import React, { useEffect, useState, useRef, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useLoop } from '../hooks';
import {
  uuid,
  linearEase,
  getWrapperStyle,
  getTimeStyle,
  svgStyle,
  getPath,
  getNormalizedColors,
  getStroke,
  colorsValidator
} from '../utils';

const getGradientId = (isLinearGradient, gradientUniqueKey) => (
  isLinearGradient ? `countdown-circle-timer-gradient-${gradientUniqueKey || uuid()}` : ''
);

const CountdownCircleTimer = props => {
  const {
    size,
    strokeWidth,
    trailColor,
    durationSeconds,
    isPlaying,
    colors,
    strokeLinecap,
    timeFormatter,
    isLinearGradient,
    gradientUniqueKey
  } = props;

  const pathRef = useRef(null);
  const [pathTotalLength, setPathTotalLength] = useState(0);
  const path = useMemo(() => getPath(size, strokeWidth), [size, strokeWidth]);
  const durationMilliseconds = useMemo(() => durationSeconds * 1000, [durationSeconds]);
  const normalizedColors = useMemo(() => getNormalizedColors(colors, durationMilliseconds, isLinearGradient), [colors, durationMilliseconds, isLinearGradient]);
  const gradientId = useMemo(() => getGradientId(isLinearGradient, gradientUniqueKey), [isLinearGradient, gradientUniqueKey]);

  useEffect(() => {
    const totalLength = pathRef.current.getTotalLength().toFixed(2);
    setPathTotalLength(totalLength);
  }, []);

  const elapsedTime = useLoop(durationMilliseconds, isPlaying);
  const strokeDasharray = linearEase(elapsedTime, 0, pathTotalLength, durationMilliseconds).toFixed(2);
  const stroke = getStroke(normalizedColors, elapsedTime);
  const remainingTime = Math.ceil((durationMilliseconds - elapsedTime) / 1000);

  return (
    <div style={getWrapperStyle(size)}>
      <svg width={size} height={size} style={svgStyle} xmlns="http://www.w3.org/2000/svg">
        {isLinearGradient && (
          <defs>
            <linearGradient id={gradientId} x1="100%" y1="0%" x2="0%" y2="0%">
              {normalizedColors.map(color => <stop {...color.gradient} />)}
            </linearGradient>
          </defs>
        )}
        <path
          fill="none"
          strokeWidth={strokeWidth}
          stroke={trailColor}
          d={path}
        />
        <path
          fill="none"
          stroke={isLinearGradient ? `url(#${gradientId})` : stroke}
          d={path}
          ref={pathRef}
          strokeLinecap={strokeLinecap}
          strokeWidth={strokeWidth}
          strokeDasharray={pathTotalLength}
          strokeDashoffset={strokeDasharray}
        />
      </svg>
      <div style={getTimeStyle(stroke, size)}>
        {typeof timeFormatter === 'function' && timeFormatter(remainingTime, elapsedTime)}
      </div>
    </div>
  );
};

CountdownCircleTimer.propTypes = {
  durationSeconds: PropTypes.number.isRequired,
  colors: PropTypes.arrayOf(
    PropTypes.arrayOf(colorsValidator).isRequired
  ).isRequired,
  size: PropTypes.number,
  strokeWidth: PropTypes.number,
  trailColor: PropTypes.string,
  isPlaying: PropTypes.bool,
  strokeLinecap: PropTypes.oneOf(['round', 'square']),
  timeFormatter: PropTypes.func,
  isLinearGradient: PropTypes.bool,
  gradientUniqueKey: PropTypes.string
};

CountdownCircleTimer.defaultProps = {
  size: 180,
  strokeWidth: 12,
  trailColor: '#d9d9d9',
  isPlaying: false,
  strokeLinecap: 'round',
  isLinearGradient: false
};

export {
  CountdownCircleTimer
};