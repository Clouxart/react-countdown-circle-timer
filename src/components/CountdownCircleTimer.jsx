import React, { useEffect, useState, useRef, useMemo } from 'react'
import PropTypes from 'prop-types'
import { useElapsedTime } from 'use-elapsed-time'
import {
  uuid,
  linearEase,
  getWrapperStyle,
  getTimeStyle,
  svgStyle,
  getPath,
  getNormalizedColors,
  getStroke,
  colorsValidator,
  visuallyHidden,
  getPathTotalLength,
} from '../utils'

const getGradientId = (isLinearGradient, gradientUniqueKey) =>
  isLinearGradient
    ? `countdown-circle-timer-gradient-${gradientUniqueKey || uuid()}`
    : ''

const getStartAt = (initialRemainingTime, startAtSeconds, durationSeconds) => {
  if (typeof initialRemainingTime === 'number') {
    return (durationSeconds - initialRemainingTime) * 1000
  }

  return startAtSeconds ? startAtSeconds * 1000 : 0
}

const CountdownCircleTimer = (props) => {
  const {
    size,
    strokeWidth,
    trailColor,
    durationSeconds,
    isPlaying,
    colors,
    strokeLinecap,
    renderTime,
    isLinearGradient,
    gradientUniqueKey,
    onComplete,
    ariaLabel,
    renderAriaTime,
    initialRemainingTime,
    startAt: startAtSeconds,
  } = props

  const pathRef = useRef(null)
  const [pathTotalLength, setPathTotalLength] = useState(0)

  // useElapsedTime will make this component rerender on every frame.
  // We memo all props that need to be computed to avoid doing that on every render
  const path = useMemo(() => getPath(size, strokeWidth), [size, strokeWidth])
  const durationMilliseconds = useMemo(() => durationSeconds * 1000, [
    durationSeconds,
  ])
  const startAt = useMemo(
    () => getStartAt(initialRemainingTime, startAtSeconds, durationSeconds),
    [initialRemainingTime, startAtSeconds, durationSeconds]
  )
  const normalizedColors = useMemo(
    () => getNormalizedColors(colors, durationMilliseconds, isLinearGradient),
    [colors, durationMilliseconds, isLinearGradient]
  )
  const gradientId = useMemo(
    () => getGradientId(isLinearGradient, gradientUniqueKey),
    [isLinearGradient, gradientUniqueKey]
  )

  useEffect(() => {
    const totalLength = getPathTotalLength(pathRef.current)
    setPathTotalLength(totalLength)
  }, [])

  const elapsedTime = useElapsedTime(isPlaying, {
    durationMilliseconds,
    onComplete,
    startAt,
  })
  const strokeDashoffset = linearEase(
    elapsedTime,
    0,
    pathTotalLength,
    durationMilliseconds
  ).toFixed(2)
  const stroke = getStroke(normalizedColors, elapsedTime)
  const remainingTime = Math.ceil((durationMilliseconds - elapsedTime) / 1000)

  return (
    <div style={getWrapperStyle(size)} aria-label={ariaLabel}>
      <svg
        width={size}
        height={size}
        style={svgStyle}
        xmlns="http://www.w3.org/2000/svg"
      >
        {isLinearGradient && (
          <defs>
            <linearGradient id={gradientId} x1="100%" y1="0%" x2="0%" y2="0%">
              {normalizedColors.map((color) => (
                <stop {...color.gradient} />
              ))}
            </linearGradient>
          </defs>
        )}
        <path
          fill="none"
          strokeWidth={strokeWidth}
          stroke={trailColor}
          d={path}
        />
        <path
          fill="none"
          stroke={isLinearGradient ? `url(#${gradientId})` : stroke}
          d={path}
          ref={pathRef}
          strokeLinecap={strokeLinecap}
          strokeWidth={pathTotalLength === 0 ? 0 : strokeWidth}
          strokeDasharray={pathTotalLength}
          strokeDashoffset={strokeDashoffset}
        />
      </svg>
      {typeof renderTime === 'function' && (
        <div aria-hidden="true" style={getTimeStyle(stroke, size)}>
          {renderTime(remainingTime, elapsedTime, isPlaying)}
        </div>
      )}
      {typeof renderAriaTime === 'function' && (
        <div role="timer" aria-live="assertive" style={visuallyHidden}>
          {renderAriaTime(remainingTime, elapsedTime, isPlaying)}
        </div>
      )}
    </div>
  )
}

CountdownCircleTimer.propTypes = {
  durationSeconds: PropTypes.number.isRequired,
  colors: PropTypes.arrayOf(PropTypes.arrayOf(colorsValidator).isRequired)
    .isRequired,
  size: PropTypes.number,
  strokeWidth: PropTypes.number,
  trailColor: PropTypes.string,
  isPlaying: PropTypes.bool,
  strokeLinecap: PropTypes.oneOf(['round', 'square']),
  renderTime: PropTypes.func,
  isLinearGradient: PropTypes.bool,
  gradientUniqueKey: PropTypes.string,
  onComplete: PropTypes.func,
  ariaLabel: PropTypes.string,
  renderAriaTime: PropTypes.func,
  initialRemainingTime: PropTypes.number,
  startAt: PropTypes.number, // To be removed in next major release v.2 use initialRemainingTime instead
}

CountdownCircleTimer.defaultProps = {
  size: 180,
  strokeWidth: 12,
  trailColor: '#d9d9d9',
  isPlaying: false,
  strokeLinecap: 'round',
  isLinearGradient: false,
  ariaLabel: 'Countdown timer',
  startAt: 0,
}

CountdownCircleTimer.displayName = 'CountdownCircleTimer'

export { CountdownCircleTimer }
