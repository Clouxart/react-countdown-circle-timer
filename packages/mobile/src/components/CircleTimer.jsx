import React from 'react'
import { View, Animated ,Dimensions} from 'react-native'
import Svg, { Path, Defs, LinearGradient, Stop, Circle } from 'react-native-svg'
import {
  DefsLinearGradient,
  countdownCircleTimerProps,
} from '@countdown-circle-timer/shared'
import { TimeWrapper } from './TimeWrapper'
import { useCountdown } from '../hooks'

const AnimatedPath = Animated.createAnimatedComponent(Path)

const windowHeight= Dimensions.get('window').height;

const countdownCircleTimerDefaultProps = {
  size:windowHeight*0.3694581,
  strokeWidth: windowHeight*0.0209359,
  trailColor: 'black',
  trailStrokeWidth:windowHeight*0.0199359,
  isPlaying: false,
  strokeLinecap: 'round',
  isLinearGradient: false,
  ariaLabel: 'Countdown timer',
  children: null,
  rotation: 'clockwise',
}

const CircleTimer = (props) => {
  const {
    size,
    strokeWidth,
    trailColor,
    trailStrokeWidth,
    duration,
    isPlaying,
    colors,
    strokeLinecap,
    children,
    isLinearGradient,
    gradientUniqueKey,
    onComplete,
    ariaLabel,
    renderAriaTime,
    initialRemainingTime,
    rotation,
  } = props

  const {
    path,
    pathLength,
    styles,
    gradientId,
    animatedElapsedTime,
    animatedStroke,
    strokeDashoffset,
    durationMilliseconds,
    isProgressPathVisible,
  } = useCountdown({
    isPlaying,
    duration,
    size,
    // https://github.com/vydimitrov/react-countdown-circle-timer/pull/82#issuecomment-774961578
    // Find the larger strokeWidth and calculate the path.
    strokeWidth: Math.max(strokeWidth, trailStrokeWidth ?? 0),
    colors,
    initialRemainingTime,
    onComplete,
    gradientUniqueKey,
    rotation,
  })

  return (
    <View
      style={styles.wrapper}
      accessible={true}
      accessibilityLabel={ariaLabel}
    >
      <Svg width={size} height={size}>
        {isLinearGradient && (
          <DefsLinearGradient
            colors={colors}
            gradientId={gradientId}
            defs={Defs}
            linearGradient={LinearGradient}
            stop={Stop}
          />
        )}
        <Path
          fill="none"
          strokeWidth={trailStrokeWidth ?? strokeWidth}
          stroke={trailColor}
          d={path}
        />
        {isProgressPathVisible && (
          <>
          <Path
      fill="none"
      stroke="black"
      d={path}
      strokeWidth="20"
      strokeDasharray="3 8"
    />
          <AnimatedPath
             fill="none"
             stroke="white"
             d={path}
             strokeWidth="22"
             strokeDasharray={pathLength}
             strokeDashoffset={strokeDashoffset}
          />
          </>
        )}
      </Svg>
      {(children !== null || typeof renderAriaTime === 'function') && (
        <TimeWrapper
          animatedColor={animatedStroke}
          animatedElapsedTime={animatedElapsedTime}
          duration={durationMilliseconds / 1000} // durationMilliseconds is locked version of the duration
          renderAriaTime={renderAriaTime}
        >
          {children}
        </TimeWrapper>
      )}
    </View>
  )
}

CircleTimer.propTypes = countdownCircleTimerProps
CircleTimer.defaultProps = countdownCircleTimerDefaultProps
CircleTimer.displayName = 'CircleTimer'

export { CircleTimer }
