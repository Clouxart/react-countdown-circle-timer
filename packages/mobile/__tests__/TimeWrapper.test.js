import React from 'react'
import { Text } from 'react-native'
import renderer from 'react-test-renderer'
import { render, act } from 'react-native-testing-library'

import { TimeWrapper } from '../src/components'

Math.random = () => 0.124578

const fixture = {
  durationMilliseconds: 10000,
  animatedElapsedTime: { addListener: jest.fn(), removeListener: jest.fn() },
  renderAriaTime: ({ remainingTime }) => remainingTime,
  animatedColor: {},
  children: ({ remainingTime }) => <Text>{remainingTime}</Text>,
}

describe('snapshot tests', () => {
  it('renders when children is function', () => {
    const tree = renderer.create(<TimeWrapper {...fixture} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders when children is component', () => {
    const TestComponent = ({ remainingTime }) => <Text>{remainingTime}</Text>
    const tree = renderer
      .create(
        <TimeWrapper {...fixture}>
          <TestComponent />
        </TimeWrapper>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('behaviour tests', () => {
  it('should pass the new remaining time to the children when new value is provided by the listener', () => {
    let listener
    const animatedElapsedTime = {
      removeListener: jest.fn(),
      addListener: (cb) => {
        listener = cb
      },
    }
    const { getByText } = render(
      <TimeWrapper
        {...fixture}
        animatedElapsedTime={animatedElapsedTime}
        renderAriaTime={undefined}
      />
    )

    act(() => {
      listener({ value: 4823 })
    })

    expect(getByText('6')).toBeTruthy()
  })
})
