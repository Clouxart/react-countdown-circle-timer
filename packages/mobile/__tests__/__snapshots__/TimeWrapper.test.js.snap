// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders when children is component 1`] = `
Array [
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      10
    </Text>
  </View>,
  <Text
    accessibilityLiveRegion="assertive"
    accessibilityRole="timer"
    importantForAccessibility="yes"
    style={
      Object {
        "height": 0,
        "opacity": 0,
      }
    }
  >
    10
  </Text>,
]
`;

exports[`snapshot tests renders when children is function 1`] = `
Array [
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      10
    </Text>
  </View>,
  <Text
    accessibilityLiveRegion="assertive"
    accessibilityRole="timer"
    importantForAccessibility="yes"
    style={
      Object {
        "height": 0,
        "opacity": 0,
      }
    }
  >
    10
  </Text>,
]
`;
